### PR TITLE
added typescript declarations

### PR DESCRIPTION
**Pull request by kaoDev**
Hey,
I used glamorous-native in a Typescript project and wanted to contribute my declaration files so that other Typescript projects can benefit.

Greetings 
Kalle

### CODE REVIEW

**Comment by ajwhite** [comment]
Thanks for this contribution! @luke-john has done a lot with Typescript over in [paypal/glamorous](https://github.com/paypal/glamorous), would you mind taking a look? I'm probably not the best to review this

**Comment by kaoDev** [comment]
yes I think @luke-john is just the right person :), I used the paypal/glamorous typings as a basis to create thos for native

File: typings/component-factory.d.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  >
+
+  <Props extends { theme?: any }>(
+    ...styles: StyleArgument<Properties, Props & ExternalProps & DefaultProps>[]
+  ): GlamorousComponent<
+    ElementProps &
+      ExternalProps &
+      Partial<DefaultProps> &
+      Omit<Props, 'theme'> &
+      Props,
```
Comment (luke-john) [inline typings/component-factory.d.ts]: This `Props` was a bug in the glamorous `typings` which has been fixed.

Comment (luke-john) [reply typings/component-factory.d.ts]: Also this is missing `ElementProps` (also recently fixed in glamorous)

**Comment by luke-john** [comment]
This is awesome, really happy to see react-native get typings as well 💖.

I wonder if we could share some of the typings across `glamorous` and `glamorous-native` 🤔 , that way both projects would get to take advantage of any improvements and fixes to the shared parts.  Reading through these typings glamorous has already received a couple of improvements since these were branched.

The easiest way would be to add glamorous as a dependency and these typings could import the common parts from there, but that's a pretty massive dependency just for the typings.

Alternatively, and probably the better solution, would be to re-publishing the typings seperate to glamorous.

I'm not likely to get time to look into getting this setup in the next week though, so I'd proceed with landing these typings as is 👍 .

**Comment by kaoDev** [comment]
Hey Luke thank you for the feedback, would be really cool to have a common base for the typings. I will try to get some time to think about this, but the next weeks I am still pretty stuffed with work

**Comment by epeli** [comment]
Hi, I'd like to test these typings but I'm having hard time figuring out how to add these to my project so that TS would pick them up for `glamorous-native` imports.

I tried creating `typings/glamorous-native` with the type definitions files and adding `tsconfig.json` entry: `"typeRoots": ["./node_modules/@types", "./typings"]` but no luck so far.

Could someone point me to right direction? Thanks.

**Comment by ajwhite** [comment]
Aiming to merge this up! @kaoDev any insight on the above if any changes are needed in this PR?

**Comment by luke-john** [comment]
Unfortunately the way typescript handles definitions under `typeRoots` is not consistant with bundled typings (or packaged types  ¯\_(ツ)_/¯).  The easiest way to give them a go for testing would be to add them into the `glamorous-native` folder in node_modules and update the `package.json` to contain the following property.

`  "types": "./typings/glamorous.d.ts",`

**Comment by kaoDev** [comment]
hi @epeli, 
TypeScript looks for the `types` entry in the package.json, in this PR it is added to point to "./typings/glamorous.d.ts" as an entrypoint. So when it is merged it should work out of the box.
I just searched for a good tutorial to configure typescript to handle custom typings for untyped node-modules but unfortunately wasn't successful. So a very short writeup, potentially incomplete:

1. add a folder "declarations" to your project
2. add a folder named like the module you are writing/testing declarations for
3. inside the module folder add the declaration files, if it is an index.d.ts TypeScript will match it automatticaly otherwise add a package.json with the "types" property pointing to the typings main file
4. add the baseUrl and paths to the tsconfig:
    ```json     
    "baseUrl": ".",
    "paths": {
          "*": [
                "declarations/*"
          ]
    }
    ```
5. make sure "moduleResolution": "node" is set in the tsconfig file

I hope this is helpful and you can test the typings 😄

**Comment by epeli** [comment]
> I just searched for a good tutorial to configure typescript to handle custom typings for untyped node-modules but unfortunately wasn't successful.

Yeah I tried that too with no luck but your writeup seems good, thanks! I'll post here if I find something strange.

**Comment by epeli** [comment]
No worries. Thanks for this!

Didn't find anything unusual during my testing btw.

**Comment by epeli** [comment]
Now I did. There is a definition `withProps` but that does not seem to actually exists glamorous-native but it does on the web based glamorous.

**Comment by kaoDev** [comment]
ah, thanks for the report. I will try to find some time today to make a PR for this bad copy pasta

**Comment by kaoDev** [comment]
https://github.com/robinpowered/glamorous-native/pull/77

**Comment by epeli** [comment]
How about usage of `propsAreStyleOverrides`?  At least at first glance it seems to be fundamentally incompatible with static typing as it changes the type by runtime mutation. 

I'm trying make an animated image

```js
const Character = g(Animated.Image)({
    height: 270,
    resizeMode: "contain",
    width: "100%",
});
Character.propsAreStyleOverrides = true; // type error
```

    Property 'propsAreStyleOverrides' does not exist on type 'GlamorousComponent<object & {}, object>'.


```js
<Character
  source={imageSource}
  top={animatedValue} // type error
```

    Type '{ source: any; top: Value; }' has no properties in common with type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<ExtraGlamorousProps & object, ComponentS...'.

I think this is workaroundable with some clever type assertion but I didn't figure out that yet.

But in the end I'd like to see some other API for this which would be better fit for typed languages (ts&flow).

**Comment by luke-john** [comment]
@epeli, assuming glamorous-native supports `propsAreCssOverrides`, you should be able to achieve it like follows currently.

I've added a comment to the issue created by @ajwhite about what would be needed to bring that typing support natively to `glamorous-native`.


```ts
const propsAreCssOverrides = {propsAreCssOverrides: true as false}
const Character = g(Animated.Image, propsAreCssOverrides)({
    height: 270,
    resizeMode: "contain",
    width: "100%",
});
```

**Comment by epeli** [comment]
Nope

    Type '{ source: any; top: Value; }' has no properties in common with type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<ExtraGlamorousProps & object, ComponentS...'.

The `Value` is `Animated.Value` btw.

**Comment by slorber** [comment]
Hey all.

I'm not using typescript, but it seems typescript definitions are useful for my Intellij/Webstorm IDE in many cases where it can't infer proper types.

In this lib I constantly get failing imports on Glamorous-native, does anyone know how to add support for it in Intellij?

image


I tried to download the typings but for this lib it does not appear in the download list (probably because they are not published?)
https://stackoverflow.com/questions/35741847/cannot-resolve-symbol-when-using-es6-import-syntax

Also tried to setup the bindings manually but I failed (not sure exactly how to do that)

**Pull request merged**